code/edgemedia/src/main/java/com/adobe/marketing/mobile/edge/media/internal/MediaXDMEventHelper.kt

```diff
@@ -0,0 +1,266 @@
+import com.adobe.marketing.mobile.edge.media.internal.xdm.XDMAdvertisingPodDetails
+import com.adobe.marketing.mobile.edge.media.internal.xdm.XDMChapterDetails
+import com.adobe.marketing.mobile.edge.media.internal.xdm.XDMCustomMetadata
+import com.adobe.marketing.mobile.edge.media.internal.xdm.XDMErrorDetails
+import com.adobe.marketing.mobile.edge.media.internal.xdm.XDMPlayerStateData
+import com.adobe.marketing.mobile.edge.media.internal.xdm.XDMQoeDataDetails
+import com.adobe.marketing.mobile.edge.media.internal.xdm.XDMSessionDetails
+import com.adobe.marketing.mobile.edge.media.internal.xdm.XDMStreamType
+import com.adobe.marketing.mobile.services.Log
+
+internal class MediaXDMEventHelper {
+    companion object {
+        private val sourceTag = "MediaXDMEventHelper"
+
+        private val standardMediaMetadataSet: Set<String> = setOf(
+            MediaConstants.VideoMetadataKeys.AD_LOAD,
+            MediaConstants.VideoMetadataKeys.ASSET_ID,
+            MediaConstants.VideoMetadataKeys.AUTHORIZED,
+            MediaConstants.VideoMetadataKeys.DAY_PART,
+            MediaConstants.VideoMetadataKeys.EPISODE,
+            MediaConstants.VideoMetadataKeys.FEED,
+            MediaConstants.VideoMetadataKeys.FIRST_AIR_DATE,
+            MediaConstants.VideoMetadataKeys.FIRST_DIGITAL_DATE,
+            MediaConstants.VideoMetadataKeys.GENRE,
+            MediaConstants.VideoMetadataKeys.MVPD,
+            MediaConstants.VideoMetadataKeys.NETWORK,
+            MediaConstants.VideoMetadataKeys.ORIGINATOR,
+            MediaConstants.VideoMetadataKeys.RATING,
+            MediaConstants.VideoMetadataKeys.SEASON,
+            MediaConstants.VideoMetadataKeys.SHOW,
+            MediaConstants.VideoMetadataKeys.SHOW_TYPE,
+            MediaConstants.VideoMetadataKeys.STREAM_FORMAT,
+            MediaConstants.AudioMetadataKeys.ALBUM,
+            MediaConstants.AudioMetadataKeys.ARTIST,
+            MediaConstants.AudioMetadataKeys.AUTHOR,
+            MediaConstants.AudioMetadataKeys.LABEL,
+            MediaConstants.AudioMetadataKeys.PUBLISHER,
+            MediaConstants.AudioMetadataKeys.STATION
+        )
+
+        private val standardAdMetadataSet: Set<String> = setOf(
+            MediaConstants.AdMetadataKeys.ADVERTISER,
+            MediaConstants.AdMetadataKeys.CAMPAIGN_ID,
+            MediaConstants.AdMetadataKeys.CREATIVE_ID,
+            MediaConstants.AdMetadataKeys.CREATIVE_URL,
+            MediaConstants.AdMetadataKeys.PLACEMENT_ID,
+            MediaConstants.AdMetadataKeys.SITE_ID
+        )
+
+        @JvmStatic
+        fun generateSessionDetails(mediaInfo: MediaInfo, metadata: Map<String, String>, forceResume: Boolean = false): XDMSessionDetails {
+            val sessionDetails = XDMSessionDetails()
+            sessionDetails.name = mediaInfo.id
+            sessionDetails.friendlyName = mediaInfo.name
+            sessionDetails.length = mediaInfo.length.toLong()
```

Comment:
Calling `Double.toLong()` will round any fractional part down to zero, thus losing any fractional length. However, from what I can tell the XDM schema defines SessionDetails media content length as an Integer, not Float. If it is common for customers to use fractional values for "length", then we should consider changing the XDM schemas to use Floats instead of Integer and then update SessionDetails to use Double. This appears to be a common pattern for the other classes where lengths of type Double are converted to Long to conform to the XDM schemas.

Reply:
This should be fine, we have double/float values in our public APIs in order to match the APIs for our previous SDKs.